### PR TITLE
Kapt classpath changes

### PR DESCRIPTION
**Pull request by gavra0**
This PR contains the following changes, all related to #KT-23880:

- support for constant tracking in sources
- stub generation does not need to generate history of classpath changes any more
- classpath is analyzed for changes from KAPT task

### CODE REVIEW

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptWithKotlincTask.kt [hunk 1]
```diff
@@ -54,7 +51,10 @@ open class KaptWithKotlincTask : KaptTask(), CompilerArgumentAwareWithInput<K2JV
         val pluginOptionsWithKapt: CompilerPluginOptions = pluginOptions.withWrappedKaptOptions(
             withApClasspath = kaptClasspath,
             changedFiles = changedFiles,
-            compiledSourcesDir = getCompiledSources())
+            classpathChanges = classpathChanges,
+            compiledSourcesDir = getCompiledSources(),
+            processIncrementally = processIncrementally
```
[the remark this answers is not in the document]
Comment (gavra0) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptWithKotlincTask.kt]: Because `KaptTask.isIncremental` keeps value of the flag (`kapt.incremental.apt`) I cannot use it here. `processIncrementally` is evaluated during task execution and indicates that we are indeed able to run incrementally. E.g. right now classpath snapshotting cannot handle added/removed classpath entries, but we can compare current-previous classpath only once we run.

File: plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/KaptOptions.kt [hunk 2]
```diff
@@ -45,12 +46,13 @@ class KaptOptions(
...
         val compiledSources: MutableList<File> = mutableListOf()
         var incrementalCache: File? = null
-        var classpathFqNamesHistory: File? = null
+        val classpathChanges: MutableList<String> = mutableListOf()
 
         var sourcesOutputDir: File? = null
         var classesOutputDir: File? = null
         var stubsOutputDir: File? = null
         var incrementalDataOutputDir: File? = null
+        var processIncrementally: Boolean = false
```
[the remark this answers is not in the document]
Comment (gavra0) [reply plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/KaptOptions.kt]: Yup, looks much better, thanks!

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptTask.kt [hunk 3]
```diff
@@ -159,8 +166,31 @@ abstract class KaptTask : ConventionTask(), TaskWithLocalState {
...
     protected fun getCompiledSources() = listOfNotNull(kotlinCompileTask.destinationDir, kotlinCompileTask.javaOutputDir)
 
-    protected fun getChangedFiles(inputs: IncrementalTaskInputs): List<File> {
-        return if (!isIncremental || !inputs.isIncremental || !getCompiledSources().all { it.exists() }) {
+    protected fun getIncrementalChanges(inputs: IncrementalTaskInputs): KaptIncrementalChanges {
+        val changedFiles = getChangedFiles(inputs)
+
+        return if (isIncremental) {
+            val classpathChanges = if (changedFiles.isEmpty()) {
+                classpath.files
```
[the remark this answers is not in the document]
Comment (gavra0) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptTask.kt]: I've rewritten this method, as it was really hard to follow what is going on.

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptTask.kt [hunk 4]
```diff
@@ -159,8 +166,31 @@ abstract class KaptTask : ConventionTask(), TaskWithLocalState {
...
+            } else {
+                classpath.files.let { cp ->
```
[the remark this answers is not in the document]
Comment (gavra0) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptTask.kt]: In case of a removed classpath entry, `ClasspathSnapshot.diff` will return early because previous and current classpath are not compatible. Right now, 2 classpath are compatible if they contain the same entries, in the same order. Handling addition/removals/swaps requires additional work.

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptWithKotlincTask.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexey-tsvetkov) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptWithKotlincTask.kt]: Makes sense, thanks for the clarification!

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/incremental/ClasspathSnapshot.kt [hunk 5]
```diff
@@ -95,10 +94,11 @@ open class ClasspathSnapshot protected constructor(
             }
         }
 
-        // We do not compute structural data for unchanged files of the current snapshot for performance reasons. That is why we
-        // update the previous snapshot as that one contains all entries.
-        computedClasspathData.putAll(previousData)
-        computedClasspathData.putAll(currentData)
+        // We do not compute structural data for unchanged files of the current snapshot for performance reasons.
+        // That is why we reuse the previous snapshot as that one contains all unchanged entries.
+        previousData.filter { !computedClasspathData.containsKey(it.key) }.forEach {
```
Comment (alexey-tsvetkov) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/incremental/ClasspathSnapshot.kt]: Minor, style: `filterTo`, destructuring parameters and `contains` operator can be used:
```
previousData.filterTo(computedClasspathData) { (key, _) -> key !in computedClasspathData }
```

**Comment by alexey-tsvetkov** [comment]
Merged manually,
@gavra0 thanks for the contribution!
